clash-lib/src/Clash/Backend.hs

```diff
@@ -61,6 +61,19 @@ data Usage
   | External Text
   -- ^ External use, field indicates the library name
 
+-- | Kind of a HDL type. Used to determine whether types need conversions in
+-- order to cross top entity boundaries.
+data HWKind
+  = PrimitiveType
+  -- ^ A type defined in an HDL spec. Usually types such as: bool, bit, ..
+  | SynonymType
+  -- ^ A user defined type that's simply a synonym for another type, very much
+  -- like a type synonym in Haskell. As long as two synonym types refer to the
+  -- same type, they can be used interchangeably. E.g., a subtype in VHDL.
```

Comment:
I would really like to see a test case for this.

That is, given:
```vhdl
subtype SubA is std_logic_vector(4 downto 0);
subtype SubB is std_logic_vector(4 downto 0):
signal a: SubA;
signal b: SubB;
signal s: std_logic_vector(4 downto 0);
```
I believe the following assignments would work:
```vhdl
a <= s;
s <= a;
b <= s;
s <= b;
```
but under the normal understanding of subtyping I'm unconvinced the following assignments are allowed:
```vhdl
a <= b;
b <= a;
```

But that's what the comment, and the rest of the code seem to suggest!

Reply:
Actually I stumbled on this by accident, I was surprised too! I'll add an explicit test if you remain unconvinced, but consider our I2C example; you'll see the following in `i2c.vhdl`:

```vhdl
  bitmaster_ds1 : entity bitmaster.bitmaster
    port map
      ( [..]
      , cmd    => cmd
      , [..]
      );
```

In this case, `cmd` is declared as:

```vhdl
signal cmd              : i2c_types.i2ccommand;
```

when you hop over to `bitmaster.vhdl`, you'll see the following instead:

```vhdl
entity bitmaster is
  port(-- clock
       [..]
       cmd    : in bitmaster_types.i2ccommand;
       [..]
end;
```

The VHDL Language Reference mentions the following about subtypes:

*  A subtype declaration does not define a new type.
* A subtype is the same type as its base type; thus, no type conversion is needed when objects of a subtype and its base type are assigned (in either direction). Also, the set of operations allowed on operands of a subtype is the same as the set of operations on its base type.

Reply:
Two notes:

* Although the VHDL Language Reference doesn't _explicitly_ say it, I think it implies that you can assign two subtypes as long as their base types are the same.
* Although GHDL doesn't complain, it might be worthwhile checking this in modelsim.

Reply:
Okay, it seems to work in Quartus as well. So we can leave this as is.